### ISSUE TITLE
OOM from leaky internal http2 class

### ISSUE THREAD

**Issue by Marcus-Rosti** [akka/akka-grpc#591]
### Versions used 

Akka-GRPC version: 0.6.1
```
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment (build 11.0.2+9-Ubuntu-3ubuntu118.10.3)
OpenJDK 64-Bit Server VM (build 11.0.2+9-Ubuntu-3ubuntu118.10.3, mixed mode, sharing)
```

### Expected Behavior

I've written a nosql database api wrapper that bidi streams requests, applies some small filtering and then streams responses.

jvm flags:
```
-XX:MaxHeapSize=256m -XX:InitialHeapSize=96m -XX:MaxMetaspaceSize=256m -XX:MetaspaceSize=128m -XX:MaxDirectMemorySize=64m -XX:+UseG1GC -XX:NativeMemoryTracking=summary
```

### Actual Behavior

I run the service with a heap of 256m and G1GC. About 8 hours the service out of memories and crashes. I can replicate the OOM locally by just no-oping the database call and returning None and then just calling my service around 3/4 million times.

### Relevant logs

The class with the largest memory footprint at OOM is:
```
akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1$OutStream
```
which takes up around 60% of that memory.

**Comment by Marcus-Rosti** [comment]
I used visualVM to to diagnose all of the memory issues and have some heap dumps

**Comment by TimMoore** [comment]
@Marcus-Rosti are you able to share the project you're using to reproduce this? Thanks for reporting!

**Comment by Marcus-Rosti** [comment]
Hmm I probably can't post the critical part but I could come up with a minimum case that demonstrates the bug if that's helpful

**Comment by TimMoore** [comment]
It would be helpful, thank you!

**Comment by Marcus-Rosti** [comment]
Here's my basic implementation of the batch endpoint that exhibited the OOM (you can imagine that database.getAttribute just returns Future.successful(Attribute.empty) or something similar.

```scala
class AttributeServiceImpl(
  namespace: String,
  database: DatabaseClient,
  readParallelism: Int,
  writeParallelism: Int
)(
  implicit ec: ExecutionContext,
  log: LoggingAdapter,
  collectorRegistry: CollectorRegistry,
  mat: Materializer
) extends AttributeService {

...
  private val readCounter = Counter.build
    .name("feature_gauge")
    .help("Feature gauge")
    .labelNames("feature")
    .register(collectorRegistry)

  private def readCounterLogger(attribute: Attribute): Unit =
    readCounter.labels(attribute.featureName).inc()
  
  private val readAttributeFlow =
    Flow[AttributeRequest]
      .map(req => ParsedRequests.apply(namespace, req))
      .async
      .mapAsyncUnordered(readParallelism)(database.getAttribute)
      .async
      .alsoTo(Sink.foreach(resp => readCounterLogger(resp.getAttribute)))

  override def getAttributes(request: AttributesRequest): Source[AttributeResponse, NotUsed] =
    Source(request.requests.toList)
      .via(readAttributeFlow)
      .withAttributes(supervisionStrategy(resumingDecider))
      .async
...
}
```

And my test suite (from a blocking java call)
```Java
  @Test
  public void getData() {
    final AttributesRequest.Builder requestBuilder = AttributesRequest.newBuilder();
    keys.forEach(key -> attributes.forEach(attribute -> requestBuilder
      .addRequests(AttributeRequest.newBuilder().setKey(key).setFeatureName(attribute))));

    final AttributesRequest requests = requestBuilder.build();

    List<Long> times = LongStream.range(0, 1000000)
      .parallel()
      .boxed().map(i -> make_request(requests, true))
      .collect(toImmutableList());
  }
```

**Comment by Marcus-Rosti** [comment]
I have this suspicion it's related to this PR: https://github.com/akka/akka-http/pull/2237

**Comment by Marcus-Rosti** [comment]
I'm not sure it's a leak strictly speaking. When I have it under load, it won't GC the stream connections because we create one client and it persists throughout every call. Then, I can stop all the tests and leave the server running, and then the jvm is happy to gc all of those indexed streams

**Comment by Marcus-Rosti** [comment]
Alright I just confirmed it with visual vm, long lived clients don't allow akka to clean up OutStreams in akka-http. 

Streams should be cleaned up "onComplete" rather than on "client close"

**Comment by Marcus-Rosti** [comment]
@TimMoore I came up with a minimal repro case: https://github.com/Marcus-Rosti/akka-grpc-sample-kubernetes-scala/pull/1

**Comment by Marcus-Rosti** [comment]
@TimMoore I built with this fix in akka-http https://github.com/akka/akka-http/pull/2585 and can confirm that this is the fix

**Comment by raboof** [comment]
Should be fixed by #625
